modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1769,7 +1770,138 @@ def sort_by(
+                # Heuristic for a "medium" df where we will include first 100 rows, and sample
+                # of remaining rows
+                if len(df) <= 1900:
+                    return np.quantile(
+                        np.concatenate(
+                            (
+                                df[columns[0]][:100].values,
+                                df[columns[0]][100:].sample(frac=0.05),
+                            )
+                        ),
+                        quantiles,
+                    )
+                return np.quantile(df[columns[0]].sample(frac=0.1), quantiles)
+
+            def pivot_func(samples):
+                """
+                Determine quantiles from the given samples.
+
+                This function takes as input the quantiles calculated over all partitions from 
+                `sample_func` defined above, and determines a final NPartitions.get() * 2 quantiles
+                to use to roughly sort the entire dataframe. It does so by collating all the samples
+                and computing NPartitions.get() * 2 quantiles for the overall set.
+                """
+                # We don't call `np.unique` on the samples, since if a quantile shows up in multiple
+                # partition's samples, this is probably an indicator of skew in the dataset, and we
+                # want our final partitions to take this into account.
+                all_pivots = np.array(samples).flatten()
+                quantiles = [
+                    i / (NPartitions.get() * 2) for i in range(NPartitions.get() * 2)
+                ]
+                overall_quantiles = np.quantile(all_pivots, quantiles)
+                overall_quantiles[0] = np.NINF
+                overall_quantiles[-1] = np.inf
+                return overall_quantiles
+
+            def split_func(df, pivots):
+                """
+                Split the given dataframe into the partitions specified by `pivots`.
+
+                This function takes as input a row-axis partition, as well as the quantiles determined
+                by the `pivot_func` defined above. It then splits the input dataframe into NPartitions.get() * 2
+                dataframes, with the elements in the i-th split belonging to the i-th partition, as determined
+                by the quantiles we're using.
+                """
+                groupby_col = np.digitize(df[columns[0]].squeeze(), pivots)
+                grouped = df.groupby(groupby_col)
+                groups = [
+                    grouped.get_group(i)
+                    if i in grouped.keys
+                    else pandas.DataFrame(columns=df.columns)
+                    for i in range(len(pivots))
+                ]
+                if not ascending:
+                    groups = groups[::-1]
+                return tuple(groups)
+
+            new_partitions = self._partition_mgr_cls.shuffle_partitions(
```

Comment:
If I understand the steps correctly, we are doing the following:
1. Sample our partitions
2. Determine the pivots for each one of our samples
3. Split the dataframe based on our pivots
4. Sort the final set of splits?

Reply:
Pretty much:
1. Sample quantiles from partitions
  * Sample each dataframe
  * Pick 2*NPartitions.get() (henceforth N) quantiles from sample
  * Return that to head node
2. Compute 2N quantiles for the entire dataset from the 2N^2 quantiles on the head node
3. Broadcast 2N quantiles to each row partition, and use them to split each partition into 2N splits
4. Collect all N splits for each of the new 2N partitions, and join them together to result in 2N partitions
5. Lazily map a sort over all 2N partitions
6. Use our `rebalance_partitions` to ensure resulting partitions are similarly sized.